pkg/controllers/collaset/pvccontrol/pvc_control.go

```diff
@@ -0,0 +1,358 @@
+/*
+Copyright 2024 The KusionStack Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pvccontrol
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appsv1alpha1 "kusionstack.io/operating/apis/apps/v1alpha1"
+	collasetutils "kusionstack.io/operating/pkg/controllers/collaset/utils"
+	"kusionstack.io/operating/pkg/controllers/utils/expectations"
+	refmanagerutil "kusionstack.io/operating/pkg/controllers/utils/refmanager"
+)
+
+type Interface interface {
+	GetFilteredPvcs(context.Context, *appsv1alpha1.CollaSet) ([]*corev1.PersistentVolumeClaim, error)
+	CreatePodPvcs(context.Context, *appsv1alpha1.CollaSet, *corev1.Pod, []*corev1.PersistentVolumeClaim) error
+	DeletePodPvcs(context.Context, *appsv1alpha1.CollaSet, *corev1.Pod, []*corev1.PersistentVolumeClaim) error
+	DeletePodUnusedPvcs(context.Context, *appsv1alpha1.CollaSet, *corev1.Pod, []*corev1.PersistentVolumeClaim) error
+	SetPvcsOwnerRef(*appsv1alpha1.CollaSet, []*corev1.PersistentVolumeClaim) ([]*corev1.PersistentVolumeClaim, error)
+	ReleasePvcsOwnerRef(*appsv1alpha1.CollaSet, []*corev1.PersistentVolumeClaim) ([]*corev1.PersistentVolumeClaim, error)
+}
+
+type RealPvcControl struct {
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+func NewRealPvcControl(client client.Client, scheme *runtime.Scheme) Interface {
+	return &RealPvcControl{
+		client: client,
+		scheme: scheme,
+	}
+}
+
+func (pc *RealPvcControl) GetFilteredPvcs(ctx context.Context, cls *appsv1alpha1.CollaSet) ([]*corev1.PersistentVolumeClaim, error) {
+	var filteredPVCs []*corev1.PersistentVolumeClaim
+	if cls.Spec.Selector.MatchLabels == nil {
+		return filteredPVCs, nil
+	}
+	selector, err := metav1.LabelSelectorAsSelector(cls.Spec.Selector)
+	if err != nil {
+		return filteredPVCs, err
+	}
+
+	// list pvcs match label selector of set
+	pvcList := &corev1.PersistentVolumeClaimList{}
+	err = pc.client.List(ctx, pvcList, &client.ListOptions{Namespace: cls.Namespace, LabelSelector: selector})
```

Comment:
I wonder whether these PVCs already in used still need to manage the ownership somehow, in case another CollaSet from same namespace takes over them and rebinds to new Pods.

Reply:
Users may configure "whenDelete=Retain" on cls, thus to retention pvcs when cls is deleted, controller will release cls's pvcs ownerReferences.

Reply:
A possible solution to maintain ownerRefs on pvcs is to release ownerRefs only when cls's deletetiontTimestamps!=nil. I'll fix solve this problem follow this solution.

Reply:
I'll fix this problem soon. This is because some PVCs don't have ownerReferences if user configure "whenDelete=Retain". As your advice, it is necessary to keep an persistent ownerReference for PVCs, unless their owner CollaSet is deleted and user want to retain them.

Reply:
fixed with https://github.com/KusionStack/operating/pull/162/commits/2dc64423e9ca74143ff5a09b6f1c2f549c09f85e

Reply:
fixed